dirty_cat/_fuzzy_join.py

```diff
@@ -0,0 +1,261 @@
+
+    if analyzer not in ["char", "word", "char_wb"]:
+        raise ValueError(
+            f"analyzer should be either 'char', 'word' or 'char_wb', got {analyzer!r}",
+        )
+
+    if match_type not in ["nearest", "radius"]:
+        raise ValueError(
+            f"match_type should be either 'nearest' or 'radius', got {match_type!r}",
+        )
+
+    if keep not in ["left", "right", "all"]:
+        raise ValueError(
+            f"keep should be either 'left', 'right' or 'all', got {keep!r}",
+        )
+
+    if len(suffixes) != 2:
+        raise ValueError(f"Invalid number of suffixes: expected 2, got {len(suffixes)}")
+    lsuffix, rsuffix = suffixes
+
+    overlap_cols = left_table_clean._info_axis.intersection(
+        right_table_clean._info_axis
+    )
+    if len(overlap_cols) > 0:
+        for i in range(len(overlap_cols)):
+            new_name_l = overlap_cols[i] + lsuffix
+            new_name_r = overlap_cols[i] + rsuffix
+            left_table_clean.rename(columns={overlap_cols[i]: new_name_l}, inplace=True)
+            right_table_clean.rename(
+                columns={overlap_cols[i]: new_name_r}, inplace=True
+            )
+            # Useful in case on[0]==on[1] and overlapping:
+            if len(on) == 2 and overlap_cols[i] in on[0]:
+                on[0] = new_name_l
+            if len(on) == 2 and overlap_cols[i] in on[1]:
+                on[1] = new_name_r
+
+    if not isinstance(on, list):
+        raise ValueError(
+            f"value {on!r} was specified for parameter 'on', "
+            "which has invalid type, expected list of column names."
+        )
+    if len(on) == 1:
+        left_col = on[0] + lsuffix
+        right_col = on[0] + rsuffix
+    elif len(on) == 2:
+        left_col = on[0]
+        right_col = on[1]
+    else:
+        raise ValueError(
+            "Expected a list with one or two elements for parameter 'on',"
+            f"received {len(on)} ({on})."
+        )
+
+    enc = CountVectorizer(analyzer=analyzer, ngram_range=ngram_range)
+    left_enc = enc.fit_transform(left_table_clean[left_col])
+    right_enc = enc.transform(right_table_clean[right_col])
+
+    left_enc = TfidfTransformer().fit_transform(left_enc)
+    right_enc = TfidfTransformer().fit_transform(right_enc)
+
+    # Find nearest neighbor using KNN :
+    neigh = NearestNeighbors(n_neighbors=1)
+    neigh.fit(right_enc)
+    distance, neighbors = neigh.kneighbors(left_enc, return_distance=True)
+    idx_closest = np.ravel(neighbors)
+
+    left_array = np.array(left_table_clean)
+    right_array = np.array(right_table_clean)
+    if match_type == "nearest":
+        joined = np.append(
+            left_array,
+            np.array([right_array[idx_closest[idr]] for idr in left_table_clean.index]),
+            axis=1,
+        )
+
+    elif match_type == "radius":
+        prec = np.zeros(left_enc.shape[0])
+        n_neigh = NearestNeighbors()
+        n_neigh.fit(right_enc)
```

Comment:
Potentially you should be able to reuse the previous instance, since `n_neighbors` can be passed to `kneighbors`.

Reply:
Thanks, this part will completely change.